codeworld-compiler/src/ParseCode.hs

```diff
@@ -0,0 +1,128 @@
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module ParseCode (checkParsedCode) where
+
+import qualified Data.ByteString as B
+import           Data.ByteString.Char8 (pack)
+import           Data.List.Split (splitOn)
+import           Data.Generics
+import           Language.Haskell.Exts
+
+checkParsedCode :: FilePath -> FilePath -> IO Bool
+checkParsedCode src err = do
+    result <- parseFile src
+    let parseList = words $ show result
+        parseresult = head parseList
+    if parseresult == "ParseFailed" then do
+        source <- readFile src
+        let sourceSplitList = splitOn "\n" source
+            errLin = errLineLocation parseList
+            errCol = errColumnLocation parseList
+        B.writeFile err (pack ("ParseError, Error at: program.hs:"
+                ++ (show $ errLin + 1) ++ ":" ++ (show errCol)
+                ++ "\n>  " ++ (sourceSplitList !! errLin)))
+        return False
+        else do
+            let outp = getErrors (getSrcSpanInfo result)
+            if outp == [] then return True
+                else do
+                    let outpString = tail $ tail $ init $ init (show outp)
+                    B.writeFile err (pack outpString)
+                    return False
+
+getSrcSpanInfo :: ParseResult (Module SrcSpanInfo) -> Module SrcSpanInfo
+getSrcSpanInfo (ParseOk x) = x
+
+getErrors :: Module SrcSpanInfo -> [String]
+getErrors m = everything (++) (mkQ [] badExps) m ++
+              everything (++) (mkQ [] badMatches) m
+              -- everything (++) (mkQ [] badInfixExps) m
+
+{-
+badInfixExps :: Exp SrcSpanInfo -> [String]
+badInfixExps (InfixApp a (App l _ _) c d) = ["Error at InfixApp: Missing &"
+        ++ "program.hs:"
+        ++ (show $ parseSrcSpanStartLine a)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn a)]
+badInfixExps (InfixApp a b c (App l _ _)) = ["Error at InfixApp: Missing &"
+        ++ "program.hs:"
+        ++ (show $ parseSrcSpanStartLine l)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn l)]
+badInfixExps (InfixApp a b c d) | not (isGoodInfixExpRhs c) = ["Error at InfixApp: program.hs:"
+        ++ (show $ parseSrcSpanStartLine a)
+        ++ ":"
+        ++ (show $ parseSrcSpanStartColumn a)]
+badInfixExps _                                              = []
+-}
+
+{-
+badExps (App x (Lit m _) (Lit m _))  = []
+badExps (App x (Lit m _) (Var m _))  = []
+-}
+badExps :: Exp SrcSpanInfo -> [String]
+badExps (App x (Var l _) e) | not (isGoodExpRhs e) = [
+        "Missing parantheses in function applicaion at: program.hs:"
+        ++ (show $ parseSrcSpanEndLine l) ++ ":"
```

Comment:
I think you should write this function:

    formatLocation :: SrcSpanInfo -> String
    formatLocation (SrcSpanInfo s _) = name ++ ":" ++ show line ++ ":" ++ show col
        where name = srcSpanFilename s
              line = srcSpanStartLine s
              col = srcSpanStartColumn s

Then you can use that for any line and column references.